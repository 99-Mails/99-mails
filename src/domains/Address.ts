import type { Domain } from "./Domain";
import type { Mail } from "./Mail";
import type { ID } from "./Shared";

export type Address = {
  id: ID;

  /**
   * Email address string <login>@<domain>
   */
  address: string;

  /**
   * <domain> part of .address.
   * Note: it's not domain name string! It's `Domain` object!
   */
  domain: Domain;

  /**
   * Key that can be used to restore the access to this address after session terminated.
   * See MutationRoot.restoreAddress
   */
  restoreKey: string;

  /**
   * List of all emails received by this address.
   * It's recommended to use Session.mails instead
   */
  mails: [Mail];
};

export type AddressID = Pick<Address, "address" | "id">;
