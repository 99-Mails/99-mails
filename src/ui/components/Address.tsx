import { useEffect } from "react";
import {
  Card,
  CardBody,
  CardHeader,
  Heading,
  Stack,
  Box,
  Flex,
  Text,
  StackDivider,
  Center,
  Tag,
  useClipboard,
  IconButton,
  Spinner,
} from "@chakra-ui/react";
import { CopyIcon } from "@chakra-ui/icons";
import useStore from "../../services/storeService";
import { useSession } from "../../services/api";

const AddressRow = ({ address, size }: { address: string; size: number }) => {
  const { onCopy, value, setValue, hasCopied } = useClipboard("");

  useEffect(() => {
    setValue(address);
  }, [address]);

  return (
    <Box>
      <Center>
        <Heading size="xs">
          <Flex gap={2} alignItems={"center"}>
            <Text fontSize="3xl">{address}</Text>
            <IconButton
              onClick={onCopy}
              colorScheme={hasCopied ? "green" : "gray"}
              aria-label="Search database"
              size={"xs"}
              icon={<CopyIcon />}
            />
            <Tag bg="gray.500" color="white">
              {size}
            </Tag>
          </Flex>
        </Heading>
      </Center>
    </Box>
  );
};

const Address = () => {
  const { sessionID } = useStore();
  const { getAddressWithSession } = useSession();

  const {
    data: addressList,
    loading,
  } = getAddressWithSession(sessionID);

  return (
    <Card bg="gray.200" border="1px" borderColor="gray.300" boxShadow="none">
      <CardHeader>
        <Heading size="md">Your temporary email(s)</Heading>
      </CardHeader>

      <CardBody>
        <Stack divider={<StackDivider />} spacing="4">
          {loading && (
            <Center>
              <Spinner />
            </Center>
          )}
          {addressList &&
            addressList.session.addresses.map((address, index) => (
              <AddressRow key={index} address={address.address} size={2} />
            ))}
        </Stack>
      </CardBody>
    </Card>
  );
};

export default Address;
