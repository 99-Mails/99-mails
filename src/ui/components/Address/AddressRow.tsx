import type { Address } from "../../../domains/Address";
import {
  Tag,
  Flex,
  Text,
  IconButton,
  useClipboard,
  forwardRef,
  IconButtonProps,
} from "@chakra-ui/react";
import { useEffect } from "react";
import useDeleteAddressFromSession from "../../../application/deleteAddressFromSession";
import { CopyIcon, DeleteIcon } from "@chakra-ui/icons";

const AddressRowButton = forwardRef<IconButtonProps, "button">((props, ref) => {
  return <IconButton ref={ref} {...props} />;
});

const AddressRowName = ({ name }: { name: string }) => {
  return <Text fontSize="3xl">{name}</Text>;
};

const AddressRowCount = ({ count }: { count: number }) => {
  return (
    <Tag bg="gray.500" color="white">
      {count}
    </Tag>
  );
};

const AddressRow = ({
  address,
  count,
}: {
  address: Pick<Address, "address" | "id">;
  count: number;
}) => {
  const { doDeleteAddress } = useDeleteAddressFromSession();
  const {
    onCopy: onClipboard,
    setValue,
    hasCopied: isClipboard,
  } = useClipboard("", { timeout: 0 });

  useEffect(() => {
    setValue(address.address);
  }, []);

  return (
    <Flex gap={2} alignItems={"center"}>
      <AddressRowName data-testid="email-address" name={address.address} />
      <AddressRowButton
        data-testid="copy-address"
        onClick={onClipboard}
        icon={<CopyIcon />}
        aria-label={""}
      />
      <AddressRowButton
        data-testid="delete-address"
        onClick={() => doDeleteAddress(address)}
        icon={<DeleteIcon />}
        aria-label={""}
      />
      <AddressRowCount data-testid="email-count" count={count} />
    </Flex>
  );
};

export default AddressRow;
