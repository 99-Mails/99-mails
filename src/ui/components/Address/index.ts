export * from "./Address";
export * from "./AddressRow";
