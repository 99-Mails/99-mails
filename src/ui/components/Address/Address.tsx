import { useEffect, useState } from "react";
import {
  Text,
  Flex,
  Stack,
  Center,
  Spinner,
  Heading,
  Card,
  CardBody,
  CardHeader,
  StackDivider,
  forwardRef,
  CardBodyProps,
} from "@chakra-ui/react";
import { useTempEmail } from "@/services/store";
import { useGetAddressWithSession } from "@/services/api";
import type { Address } from "@/domains/Address";
import AddressRow from "./AddressRow";
import ErrorBoundary, { ErrorFallback } from "../ErrorBoundary";
import { ApolloError } from "@apollo/client";
import { Timer } from "@/ui/components/Timer";

const AddressHeader = () => {
  return (
    <CardHeader>
      <Heading size="md">Your temporary email(s)</Heading>
    </CardHeader>
  );
};

type AddressBodyProps = CardBodyProps & {
  address: Address[];
  error: ApolloError;
  loading: boolean;
};

const AddressBody = forwardRef<AddressBodyProps, "div">((props, ref) => {
  const { address, error, loading } = props;

  return (
    <CardBody ref={ref}>
      <Stack divider={<StackDivider />} spacing="2">
        {!address && error && (
          <Center>
            <Text>Error!</Text>
          </Center>
        )}
        {loading && (
          <Center>
            <Spinner data-testid="loading-spinner" />
          </Center>
        )}
        {address.length == 0 && !loading && (
          <Center>
            <Text data-testid="no-address" as="b">
              No Active Address!
            </Text>
          </Center>
        )}
        {address &&
          address.map((address: Address, index: number) => (
            <AddressRow key={address.id} address={address} count={2} />
          ))}
      </Stack>
    </CardBody>
  );
});

const AddressWrapper = ({
  children,
  reset,
}: {
  children: React.ReactElement[] | React.ReactElement;
  reset: () => void;
}) => {
  return (
    <ErrorBoundary FallbackComponent={ErrorFallback} onReset={() => reset()}>
      <Card bg="gray.200" border="1px" borderColor="gray.300" boxShadow="none">
        {children}
      </Card>
    </ErrorBoundary>
  );
};

const AddressContainer = () => {
  const { sessionID } = useTempEmail();
  const { data, loading, refetch, error } = useGetAddressWithSession(sessionID);
  const [address, setAddress] = useState([]);

  useEffect(() => {
    function handleAddressNotLoading() {
      if (!loading && data) {
        setAddress(data?.session?.addresses);
      }
    }
    handleAddressNotLoading();

    return () => setAddress([]);
  }, [loading, data]);

  return (
    <AddressWrapper reset={refetch}>
      <AddressHeader />
      <AddressBody
        data-testid="address-list"
        address={address}
        error={error}
        loading={loading}
      />
    </AddressWrapper>
  );
};

export default AddressContainer;
