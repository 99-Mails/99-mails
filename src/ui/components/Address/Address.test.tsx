import { vi, it, expect } from "vitest";
import { 
  screen, 
  renderWithContext, 
  waitForElementToBeRemoved 
} from "../../../tests/test-utils";
import Address from "./Address";
import GetAddressWithSessionResponse from '../../../mocks/responses/GetAddressWithSession.json'

let storeProps;

beforeEach(() => {
  storeProps = {
    sessionId: "12345678",
    cleanState: vi.fn(),
    setSession: vi.fn(),
  };
});

it("should render correctly", () => {
  const { asFragment } = renderWithContext(<Address />, { storeProps });
  expect(asFragment()).toMatchSnapshot();
});

it("should display the loading spinner", () => {
  renderWithContext(<Address />, { storeProps });
  expect(screen.getByTestId("loading-spinner")).toHaveTextContent("Loading...");
});

it("should display the emails", async () => {
  renderWithContext(<Address />, { storeProps });

  const addresses = GetAddressWithSessionResponse.session.addresses;

  await waitForElementToBeRemoved(screen.queryByText('Loading...'))

  addresses.forEach(address => {
    expect(screen.getByText(address.address)).toBeDefined()
  })
});
