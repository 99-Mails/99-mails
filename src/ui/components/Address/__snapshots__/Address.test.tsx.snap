// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-cy3ym0"
  >
    <div
      class="css-gmuwbf"
    >
      <h2
        class="chakra-heading css-18j379d"
      >
        Your temporary email address
      </h2>
    </div>
    <div
      class="css-ir7jip"
      data-testid="address-list"
    >
      <div
        class="chakra-stack css-j7qwjs"
      >
        <div
          class="css-gmuwbf"
        >
          <div
            class="chakra-spinner css-nkpswe"
            data-testid="loading-spinner"
          >
            <span
              class="css-8b45rq"
            >
              Loading...
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
